core/rio/hdt/src/main/java/org/eclipse/rdf4j/rio/hdt/HDTDictionarySectionPFC.java

```diff
@@ -181,4 +252,76 @@ protected void parse(InputStream is) throws IOException {
 		}
 		return arr;
 	}
+
+	/**
+	 * Encode a single block, removing the input strings to conserve memory
+	 * 
+	 * @param iter
+	 * @return encoded block
+	 */
+	private byte[] encodeBlock(Iterator<String> iter) {
+		byte[][] tmp = new byte[stringsBlock][];
+		int i = 0;
+
+		byte[] base = iter.next().getBytes(StandardCharsets.UTF_8);
+		iter.remove();
+		// copy base string into first entry, with a trailing NULL
+		tmp[i] = new byte[base.length + 1];
+		System.arraycopy(base, 0, tmp[i], 0, base.length);
+
+		byte[] prev = base;
+
+		// encode a block, with a maximum of strings per block (minus the first base string)
+		for (i = 1; iter.hasNext() && i < stringsBlock; i++) {
+			byte[] str = iter.next().getBytes(StandardCharsets.UTF_8);
+			iter.remove();
+
+			// remove the common part, and only store the offset and the suffix and trailing NULL
+			int common = mismatch(prev, str);
+			byte[] c = VByte.encode(common);
+			tmp[i] = new byte[c.length + str.length - common + 1];
+			System.arraycopy(c, 0, tmp[i], 0, c.length);
+			System.arraycopy(str, common, tmp[i], c.length, str.length - common);
+
+			prev = str;
+		}
+
+		if (i == 1) {
+			return tmp[0];
+		}
+
+		// flatten 2-dimensional byte array to a single byte array
+		int len = 0;
+		for (int j = 0; j < i; j++) {
+			len += tmp[j].length;
+		}
+
+		byte[] ret = new byte[len];
+		for (int j = 0, idx = 0; j < i; j++) {
+			System.arraycopy(tmp[j], 0, ret, idx, tmp[j].length);
+			idx += tmp[j].length;
+		}
+		return ret;
+	}
+
+	/**
+	 * Compare 2 byte arrays and return the position of the first different byte.
+	 * 
+	 * Could be replaced with Arrays.mismatch in JDK9+
```

Comment:
That would probably be a lot faster too. Since it would probably use intrinsics and SSE instructions.

Reply:
This would actually be a candidate for a multi Java version JAR file... but not something we can do in JDK8